bson/bsoncodec.go

```diff
@@ -72,202 +72,33 @@ func (vde ValueDecoderError) Error() string {
-}
-
-// UseLocalTimeZone causes the Decoder to unmarshal time.Time values in the local timezone instead
-// of the UTC timezone.
-//
-// Deprecated: Use [go.mongodb.org/mongo-driver/bson.Decoder.UseLocalTimeZone] instead.
-func (dc *DecodeContext) UseLocalTimeZone() {
-	dc.useLocalTimeZone = true
-}
-
-// ZeroMaps causes the Decoder to delete any existing values from Go maps in the destination value
-// passed to Decode before unmarshaling BSON documents into them.
-//
-// Deprecated: Use [go.mongodb.org/mongo-driver/bson.Decoder.ZeroMaps] instead.
-func (dc *DecodeContext) ZeroMaps() {
-	dc.zeroMaps = true
-}
-
-// ZeroStructs causes the Decoder to delete any existing values from Go structs in the destination
-// value passed to Decode before unmarshaling BSON documents into them.
-//
-// Deprecated: Use [go.mongodb.org/mongo-driver/bson.Decoder.ZeroStructs] instead.
-func (dc *DecodeContext) ZeroStructs() {
-	dc.zeroStructs = true
-}
-
-// DefaultDocumentM causes the Decoder to always unmarshal documents into the M type. This
-// behavior is restricted to data typed as "interface{}" or "map[string]interface{}".
-//
-// Deprecated: Use [go.mongodb.org/mongo-driver/bson.Decoder.DefaultDocumentM] instead.
-func (dc *DecodeContext) DefaultDocumentM() {
-	dc.defaultDocumentType = reflect.TypeOf(M{})
-}
-
-// DefaultDocumentD causes the Decoder to always unmarshal documents into the D type. This
-// behavior is restricted to data typed as "interface{}" or "map[string]interface{}".
-//
-// Deprecated: Use [go.mongodb.org/mongo-driver/bson.Decoder.DefaultDocumentD] instead.
-func (dc *DecodeContext) DefaultDocumentD() {
-	dc.defaultDocumentType = reflect.TypeOf(D{})
-}
-
-// ValueCodec is an interface for encoding and decoding a reflect.Value.
-// values.
-//
-// Deprecated: Use [ValueEncoder] and [ValueDecoder] instead.
-type ValueCodec interface {
-	ValueEncoder
-	ValueDecoder
+// EncoderRegistry is an interface provides a ValueEncoder based on the given reflect.Type.
+type EncoderRegistry interface {
+	LookupEncoder(reflect.Type) (ValueEncoder, error)
 }
 
 // ValueEncoder is the interface implemented by types that can encode a provided Go type to BSON.
 // The value to encode is provided as a reflect.Value and a bson.ValueWriter is used within the
 // EncodeValue method to actually create the BSON representation. For convenience, ValueEncoderFunc
-// is provided to allow use of a function with the correct signature as a ValueEncoder. An
-// EncodeContext instance is provided to allow implementations to lookup further ValueEncoders and
-// to provide configuration information.
+// is provided to allow use of a function with the correct signature as a ValueEncoder. A pointer
+// to a Registry instance is provided to allow implementations to lookup further ValueEncoders.
 type ValueEncoder interface {
-	EncodeValue(EncodeContext, ValueWriter, reflect.Value) error
+	EncodeValue(EncoderRegistry, ValueWriter, reflect.Value) error
```

Comment:
We don't want to remove `EncodeContext` from the `ValueEncoder` API, only from the `bson.Encoder` API. Keeping `EncodeContext` in the `EncodeValue` function may significantly simplify this PR.

Reply:
Current `EncodeContext`/`DecodeContext` is coupled with the fields in `Encoder`/`Decoder` and codecs. For example, `OverwriteDuplicatedInlinedFields` exists in [StructCodec](https://github.com/mongodb/mongo-go-driver/blob/v1/bson/bsoncodec/struct_codec.go#L114), [EncodeContext](https://github.com/mongodb/mongo-go-driver/blob/v1/bson/bsoncodec/bsoncodec.go#L200), and [Encoder](https://github.com/mongodb/mongo-go-driver/blob/master/bson/encoder.go#L29). Moreover, the example here, `OverwriteDuplicatedInlinedFields`, is default true. Considering future non-zero default codec fields as well as the current use cases of codec configurations for `mgocompat`, using an empty struct, such as `type structCodec struct {}`, and passing `EncodeContext`/`DecodeContext` to modify the encoding/decoding will be similarly complex. Meanwhile, the `*Context` becomes a superset of all codec attributes since it has to be passed into different codecs. Also for custom codecs, `*Context` is either hard to be extended or useless.

Though `SetBehavior` and `Registry.SetCodecOption` are out of the scope of GODRIVER-2714, they also decouple the logic with individual codecs by distributing the configuration response to each codec, yet providing a single entry from outside.

However, I agree that `SetBehavior` and `Registry.SetCodecOption` are redundant. We can make `Registry.SetCodecOption` accessible to the `bson` package only.

A less radical PR keeping both `*Context` and codec fields is available at #1659.